README.md
# Proof-of-concept Harvesting Application

This is a proof-of-concept application for harvesting of data from [Open Data Node](http://opendatanode.org/) via API showing how to work with relational datasets published using [uv-t-relationalDiffToCkan DPU](https://github.com/OpenDataNode/UVPlugin-relationalDiffToCKAN).

# How it works

This is proof-of-concept which:

1. simulates some dataset which is being updated gradually over time
2. provides example application which can download a copy that dataset and maintain this copy up-to-date

## Simulated dataset

'[t-generatedToRelational DPU](./t-generatedToRelational/)' is used in pipeline in ODN/UnifiedViews (schedulted to run every 15 minutes) to generate some dummy data:

| id  | data                |
| --- | ------------------- |
| 1   | 2016-02-29 23:03:43 |
| 2   | 2016-03-01 12:03:44 |

In short: Each day new item is added. EAch 15 minutes it gets updated (putting current date and time into 'data' column). The next day item gets deleted. That represents some quite typical datasets hich do change in time but do not track changes being done to them.

'uv-t-relationalDiffToCkan DPU' is then used to detect the differences/updates in that data and load it into ODN catalog, adding columns 'modified_timestamp' and 'deleted_timestamp' in the process to mark new, modified and deleted items:

| id  | data                | modified_timestamp  | deleted_timestamp   |
| --- | ------------------- | ------------------- | ------------------- |
| 1   | 2016-02-29 23:03:43 | 2016-03-01T00:03:47 | 2016-03-01T00:03:47 |
| 2   | 2016-03-01 12:03:44 | 2016-03-01T12:03:44 |                     |

So the purpose of 'uv-t-relationalDiffToCkan' is to add change tracking missing in original dataset:

- changed items gets their 'modified_timestamp' column update to data and time when change was detected
- deleted items are peserved, and both 'modified_timestamp' and 'deleted_timestamp' are updated (first one, because deletion is change too, second to indicate that the update was deletion)

You can see a result of that as dataset 'gen-data' at http://data.comsode.eu/dataset/gen-data .

## Harvesting application

The purpose of the harvesting application is to:

1. on first run: to get whole copy of the dataset to your PC
2. on subsequent runs: to download only changed items

Such harvesting is intended mainly for bigger datasets to avoid repeated download of items which have not changed since last harvesting.

Note: This application is written mainly for easy code reading, effective operation is only second goal. So by tweaking it, you might get better performance.

# How to use it

Simply:

1. checkout the app from GitHub
2. go into 'test_app' subdirectory
3. run the harvester: `python read_rest.py -u http://data.comsode.eu/api/action/ -i 210948d1-087f-4bb4-bc54-6f3205c4eefa -f /tmp/records.csv`

## More information

How to run harvesting application:

`python read_rest.py -u {host}/api/action/ -i {resource_id} -f {local file}`

Harvesting application has 4 parameters, 2 are mandatory.

- mandatory parameter `-i` followed by CKAN datastore resource ID to process
- parameter `-u` followed by URL of CKAN installation API, where the dataset was created (default value is `http://localhost/api/action/`)
- parameter `-a` followed by user API KEY, it is required to access resources of private datasets
- mandatory parameter '-f' followed by path to CSV file, where the records will be saved

First run of the script will download all rows of CKAN datastore resource and save them to CSV file specified in -f option. Subsequent runs will download and save only records which changed since last run.

Note: Internal CKAN IDs (column '_id') can be seen on catalogue but are skipped in local copy of data.

# Licensing

This proof-of-concept code is licensed as follows:

- t-generatedToRelational DPU is licensed under LGPLv3 (see [LICENSE](./t-generatedToRelational/LICENSE) file)
- harvesting application is licensed under Simplified BSD license (see [LICENSE](./test_app/LICENSE) file)


t-generatedToRelational/CHANGELOG.md
T-GeneratedToRelational
-----------------------

v1.0.0-SNAPSHOT
---
* initial release

test_app/read_rest.py
#!/usr/bin/env python
'''
Copyright (c) 2015, Martin Virag.
Copyright (c) 2015, Peter Golian.

All rights reserved.

Redistribution and use in source and binary forms, with or without
modification, are permitted provided that the following conditions are met:

* Redistributions of source code must retain the above copyright notice, this
  list of conditions and the following disclaimer.

* Redistributions in binary form must reproduce the above copyright notice,
  this list of conditions and the following disclaimer in the documentation
  and/or other materials provided with the distribution.

THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE
DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE LIABLE
FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR
SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER
CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY,
OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.


Simple CKAN datastore consumer example
On first run, it downloads all rows from CKAN datastore resource and saves them to
CSV file. On subsequent runs it downloads only records changed from last run.

Example of CKAN datastore resource used in this simple program:
id | data                | modified_timestamp  | deleted_timestamp
--------------------------------------------------------------------
0  | 2015-10-05 23:00:08 | 2015-10-06T00:00:10 | 2015-10-06T00:00:10
1  | 2015-10-06 13:00:28 | 2015-10-06T13:00:29 |

This example should work the same at least up to CKAN version 2.3.2

There can be little differences in API between different CKAN versions, latest:
CKAN API documentation: http://docs.ckan.org/en/latest/api/index.html
Datastore CKAN API docs: http://docs.ckan.org/en/latest/maintaining/datastore.html#the-datastore-api

You can check CKAN version with request to API: <host>/api/action/status_show

Run command:
python read_rest.py -u {host}/api/action/ -i {resource_id} -f /tmp/records.csv
'''
import csv
import optparse
import os
import json
import urllib, urllib2


def parse_args():
    usage = "python read_test.py [options]"
    parser = optparse.OptionParser(usage=usage)
    parser.add_option('-i', '--id', dest='resource_id', default=None, help='resource id (required)')
    parser.add_option('-u', '--url', dest='url', default='http://localhost/api/action/', help='URL of REST API (default: http://localhost/api/action/)')
    parser.add_option('-a', '--api_key', dest='api_key', default=None, help='API KEY is needed to access resources of private datasets')
    parser.add_option('-f', '--file', dest='file', default=None, help='Path to csv file, where the records will be saved (required)')

    (options, args) = parser.parse_args()
    # simple check for mandatory arguments
    if not options.resource_id or not options.file:
        parser.print_help()
        raise SystemExit
    
    return options, args

def send_request_to_CKAN(parameters_dict, url, api_key=None):
    request = urllib2.Request(url)
    # convert dictionary to encoded JSON string
    if parameters_dict:
        data_string = urllib.quote(json.dumps(parameters_dict))
    # accessing private dataset requires user's API KEY
    # for accessing public dataset it's not necessary
    if api_key:
        request.add_header('Authorization', api_key)
    try:
        response = None
        # Make the HTTP request.
        response = urllib2.urlopen(request, data_string)
        # Use the json module to load CKAN's response into a dictionary.
        response_dict = json.loads(response.read())
        return response_dict['result']
    except urllib2.HTTPError, e:
        print e.read() # some http errors return also JSON (see CKAN API doc)
        raise SystemExit
    finally:
        if response:
            response.close()

def print_records(records):
    for record in records:
        print u"id={0}; data={1}; modified_date={2}; deleted_date={3}"\
            .format(record['id'], record['data'], record['modified_timestamp'],
                    record['deleted_timestamp'])

class ResourceRecords:
    '''
    Class responsible for reading records from CSV file
    and updating the records in this file
    '''
    
    records = []
    fields = ['id', 'data', 'modified_timestamp', 'deleted_timestamp']
    last_update = None
    
    def __init__(self, csv_file_path):
        self.file_path = csv_file_path
    
    def update_records(self, update_records):
        for record in update_records:
            old_record = None
            for r in self.records:
                # warning: CKAN supports compound primary keys
                if r['id'] == record['id']: # found match
                    old_record = r
                    break
            
            if old_record: # updating records
                old_record.update(record)
            else: # adding new record
                self.records.append(record)
    
    def load_from_csv(self):
        if not os.path.exists(self.file_path):
            # creating empty csv
            file(self.file_path, "w").close()
            return
        
        # loading records from csv file
        with open(self.file_path, 'rb') as csvfile:
            reader = csv.DictReader(csvfile, delimiter=';')
            for row in reader:
                self.records.append(row)
                # getting last modification date
                if not self.last_update or self.last_update < row['modified_timestamp']:
                    self.last_update = row['modified_timestamp']

    def write_to_csv(self):
        # writing records to csv file
        with open(self.file_path, 'wb') as csvfile:
            writer = csv.DictWriter(csvfile, delimiter=';', fieldnames=self.fields)
            writer.writeheader()
            for record in self.records:
                writer.writerow(record)

def get_ids_for_sql(sql_query, options):
    '''
    Calls datastore_search_sql API call with query given and returns only ids
    '''
    print 'sql query: {}'.format(sql_query)
    parameters_dict = {}
    parameters_dict['sql'] = sql_query
    # datastore_search_sql: paging can be done through sql
    records = send_request_to_CKAN(parameters_dict,
                                   options.url + 'datastore_search_sql',
                                   options.api_key)
    # retrieve only ids from JSON response
    return [r['id'] for r in records['records']]

def update_saved_records(saved_records, record_ids, options):
    '''
    Calls datastore_search for given record_ids and updates saved_records with
    retrieved records (the updated records aren't saved to CSV file yet, 
    saved_records.write_to_csv() must be called after this method)
    '''
    if not record_ids:
        print "No records were updated"
        raise SystemExit
    
    # datastore_search_sql returns all records matching the sql query
    # datastore_search needs to be done through paging
    LIMIT = 1000 # limit records per page/request, CKAN default value: 100
    offset = 0
    total = len(record_ids)
        
    # paginate through returned ids
    while offset < total:
        print "records: {0} - {1}".format(offset, offset + LIMIT - 1)
        parameters_dict = {}
        # identify CKAN datastore resource (mandatory)
        parameters_dict['resource_id'] = options.resource_id
        # filter only selected IDs
        parameters_dict['filters'] = {'id': record_ids}
        # we want only fields (columns) saved in csv
        parameters_dict['fields'] = saved_records.fields
        # paging parameters
        parameters_dict['limit'] = LIMIT
        parameters_dict['offset'] = offset
        # calling datastore_search API call
        response = send_request_to_CKAN(parameters_dict,
                                        options.url + 'datastore_search',
                                        options.api_key)
        
        print_records(response['records'])
        # update already saved records
        saved_records.update_records(response['records'])
        offset += LIMIT

if __name__ == '__main__':
    (options, args) = parse_args()

    saved_records = ResourceRecords(options.file)
    # load previously saved records
    saved_records.load_from_csv()
    # it would be better save last_update time to DB / file, determine it
    # from periodicity or use other way
    # for simplicity, it's read from CSV as latest modified_timestamp

    if not saved_records.last_update: # first run, there are no records saved yet
        print u"requesting all records for resource: {}".format(options.resource_id)
        sql_query = u'SELECT id FROM "{0}"'.format(options.resource_id)
        record_ids = get_ids_for_sql(sql_query, options)
        update_saved_records(saved_records, record_ids, options)
        
    else: # subsequent runs, request only changes
        print u"requesting records updated from {}".format(saved_records.last_update)
        sql_query = u'SELECT id FROM "{0}" WHERE modified_timestamp > \'{1}\''.format(options.resource_id, saved_records.last_update)
        record_ids = get_ids_for_sql(sql_query, options)
        print 'changed records from last request:'
        update_saved_records(saved_records, record_ids, options)
    
    print 'saving records to CSV'
    saved_records.write_to_csv()
